src/OrleansServiceBus/Providers/Streams/EventHub/EventHubPartitionCheckpoint.cs

```diff
@@ -0,0 +1,74 @@
+﻿
+using System;
+using System.Threading.Tasks;
+using Microsoft.ServiceBus.Messaging;
+using Orleans.AzureUtils;
+
+namespace Orleans.ServiceBus.Providers
+{
+    internal class EventHubPartitionCheckpoint
+    {
+        private readonly AzureTableDataManager<EventHubPartitionCheckpointEntity> dataManager;
+        private readonly TimeSpan persistCadence;
+
+        private EventHubPartitionCheckpointEntity entity;
+        private Task pendingSave;
+        private DateTime? nextSave;
+
+        public bool Exists { get { return entity != null && entity.Offset != EventHubConsumerGroup.StartOfStream; } }
+
+        public EventHubPartitionCheckpoint(ICheckpointSettings settings, string streamProviderName, string partition)
+        {
+            if (settings == null)
+            {
+                throw new ArgumentNullException("settings");
+            }
+            if (string.IsNullOrWhiteSpace(streamProviderName))
+            {
+                throw new ArgumentNullException("streamProviderName");
+            }
+            if (string.IsNullOrWhiteSpace(partition))
+            {
+                throw new ArgumentNullException("partition");
+            }
+            persistCadence = settings.PersistCadence;
+            dataManager = new AzureTableDataManager<EventHubPartitionCheckpointEntity>(settings.TableName, settings.DataConnectionString);
+            entity = EventHubPartitionCheckpointEntity.Create(streamProviderName, settings.DeploymentId, partition);
+        }
+
+        public Task InitTableAsync()
+        {
+            return dataManager.InitTableAsync();
+        }
+
+        public async Task<string> Load()
+        {
+            Tuple<EventHubPartitionCheckpointEntity, string> results =
+                await dataManager.ReadSingleTableEntryAsync(entity.PartitionKey, entity.RowKey);
+            if (results != null)
+            {
+                entity = results.Item1;
+            }
+            return entity.Offset;
+        }
+
+        public void Update(string offset, DateTime utcNow)
+        {
+            // if offset has not changed, do nothing
+            if (string.Compare(entity.Offset, offset, StringComparison.InvariantCulture)==0)
+            {
+                return;
+            }
+
+            // if we've saved before but it's not time for another save or the last save operation has not completed, do nothing
+            if (nextSave.HasValue && (nextSave.Value > utcNow || !pendingSave.IsCompleted))
+            {
+                return;
+            }
+
+            entity.Offset = offset;
+            nextSave = utcNow + persistCadence ;
+            pendingSave = dataManager.UpsertTableEntryAsync(entity);
+        }
```

Comment:
nobody observes `dataManager.UpsertTableEntryAsync(entity)`. Can it throw? If it can, make sure the exception is observed (using `.Ignore()) or something like that.